src/persist-client/src/operators/shard_source.rs

```diff
@@ -0,0 +1,467 @@
+            // contain only times greater or equal to `until`, which means they can be dropped
+            // in their entirety. The current batch must be emitted, but we can stop afterwards.
+            if PartialOrder::less_equal(&until, &progress) {
+                done = true;
+            }
+
+            let parts_size_bytes = parts.iter().map(|x| x.encoded_size_bytes()).sum();
+            yield (parts, progress, parts_size_bytes);
+        }
+
+        // Rather than simply end, we spawn a task that can continue to return leases.
+        // This keeps the `ReadHandle` alive until we have confirmed each lease as read.
+        mz_ore::task::spawn(|| "LeaseReturner", async move {
+            while let Some(leased_part) = consumed_part_rx.recv().await {
+                subscription
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+        });
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_descs({})", name), scope.clone());
+    let mut flow_control_input = builder.new_input(flow_control_input, Pipeline);
+    let (mut descs_output, descs_stream) = builder.new_output();
+
+    let shutdown_button = builder.build(move |caps| async move {
+        let mut cap_set = CapabilitySet::from_elem(caps.into_element());
+
+        let mut current_ts = timely::progress::Timestamp::minimum();
+        let mut inflight_bytes = 0;
+        let mut inflight_parts: Vec<(Antichain<T>, usize)> = Vec::new();
+
+        loop {
+            // While we have budget left for fetching more parts, read from the
+            // subscription and pass them on.
+            while inflight_bytes < flow_control_max_inflight_bytes {
+                match pinned_stream.next().await {
+                    Some(Ok((parts, progress, size_in_bytes))) => {
+                        let session_cap = cap_set.delayed(&current_ts);
+                        let mut descs_output = descs_output.activate();
+                        let mut descs_session = descs_output.session(&session_cap);
+
+                        if size_in_bytes > 0 {
+                            inflight_parts.push((progress.clone(), size_in_bytes));
+                            inflight_bytes += size_in_bytes;
+                            trace!(
+                                "shard {} putting {} bytes inflight. total: {}. batch frontier {:?}",
+                                shard_id,
+                                size_in_bytes,
+                                inflight_bytes,
+                                progress,
+                            );
+                        }
+
+                        for part_desc in parts {
+                            // Give the part to a random worker.
+                            let worker_idx = usize::cast_from(Instant::now().hashed()) % num_workers;
```

Comment:
I think this can be done in a simpler and fairer way (randomness can have hotspots) with this:

```rust
// define this at the top
let worker_round_robin = (0..num_workers).cycle()

let worker_idx = worker_round_robin.next().expect("infinite iterator");
```

Reply:
Hm, I'm not sure in what context this suggestion was suggested but this exchange can produce big skews. `Instant::now()` is not a good substitute for a uniform distribution, we're calling in a tight loop and it might produce the same value multiple times in a row.

The round robin solution will always distribute work equally among the workers and it's simple to convince oneself that this is the case.

@frankmcsherry could you type check that the distribution produced by round-robining parts to workers is uniform and that it doesn't suffer from the skew issue that `Instant::now()` modding can fall into?

Reply:
The distribution produced by round robin has the skew issues that round robin might have: if your parts alternate size large, small, then you'll end up only using half of your workers. This might be fine, but a hash is probably better. You don't have to hash the time, you can just hash the round robin index, though.

Reply:
Ah, but defeating patterns cannot be done with deterministic hashing. The worker assignment produced by (hash(0), hash(1),..) is just as arbitrary and I could construct a malicious input that has a big batch every time that sequence is worker 0. I guess that was the intent of using `Instant::now()` in the first place, which has some amount of randomness.

I tested calling `Instant::now()` in a loop in release mode and it seems to produce different numbers so let's keep it. @dan I would love a comment explaining why we'd like to have some randomness in the selection

Reply:
> I would love a comment explaining why we'd like to have some randomness in the selection

done!